Throw Anonymous Child of AnalysisException [databricks]

razajafri opened the pull request:
In Spark 4.0.0 the constructor of AnalysisException that takes a String parameter has been protected. One way to get around this is to add a shim where we can throw a defined class for Spark 4.0.0 while keeping the same logic for other versions of Spark. That would be a lot of code churn and be hard to review for no added benefit IMO.

This change is small enough and simple enough to where we can throw an anonymous class without making it difficult to read

contributes to #9259 

<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/RapidsHiveErrors.scala:
```diff
@@ -54,7 +54,7 @@ object RapidsHiveErrors {
   def cannotResolveAttributeError(name: String, outputStr: String): Throwable = {
     new AnalysisException(
-      s"Unable to resolve $name given [$outputStr]")
+      s"Unable to resolve $name given [$outputStr]") {}
```
I don't think this is a good approach, as it's very unfriendly to the end-user.  The exception is going to show up without AnalysisException in the name -- it will be something like RapidsHiveErrors$$anon$## (or whatever the enclosing class is where the anonymous class was created), and thus won't look at all like the AnalysisException class that the user is used to seeing.  Yes, it's semantically like an AnalysisException since it derives from it, but the way it appears to the end user in a stacktrace does not look similar at all.

The reason Spark made the string private is because they want the code to go through QueryExecutionErrors and other similar error reporting interfaces.  We have our own wrappers to that in shims, so maybe the best way to fix this is to migrate these use cases to those already shim'd error interfaces, and have the older shims call AnalysisException and the new shims use the newer error reporting that doesn't use an arbitrary string.  If we really want to match the Spark behavior here, that's the right way to do it.

razajafri replied:
I agree, that's the right way to do it and went down that path but I wasn't sure if we will be OK with the difficulty in code review and decided to post this before going back that route. Will update in a few

razajafri commented:
@jlowe I think I have addressed all your concerns

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDataWritingCommandExec.scala:
```diff
@@ -84,10 +84,8 @@ object GpuDataWritingCommand {
-          s"CREATE-TABLE-AS-SELECT cannot create table with location to a non-empty directory " +
-              s"${tablePath} . To allow overwriting the existing non-empty directory, " +
-              s"set '$allowNonEmptyLocationInCTASKey' to true.")
+        AnalysisExceptionShim.throwException("_LEGACY_ERROR_TEMP_1241",
```
We should not be throwing literal magic strings.  This is really fragile and likely to have a typo that won't be caught by the compiler and thus likely to not do what we want at runtime.  The Spark code is not doing this in all the various places in Catalyst, instead those are calling QueryExecutionErrors with a descriptive method name, and _that_ is doing the specifics of throwing a specific message.

That's what I mentioned before:
> We have our own wrappers to that in shims, so maybe the best way to fix this is to migrate these use cases to those already shim'd error interfaces

We shouldn't be creating a new shim class here, but rather using the one we already have, i.e.: RapidsErrorUtils.  For example, for this case here we should add a `createTableAsSelectWithNonEmptyDirectoryError` method to the RapidsErrorUtils shim class.  On modern Spark versions, it would simply delegate to Spark's QueryCompilationErrors.  On older Spark versions, the RapidsErrorUtils shim would throw an AnalysisException with an appropriate message string it constructs based on the arguments.

In short, see what the new Spark code is doing in these areas.  I suspect none of them are throwing AnalysisException but rather calling an error utility class like QueryExecutionErrors or QueryCompilationErrors.  We should call the same methods on Spark versions that have those methods available, not throw a derived class.

razajafri replied:
Ah, I was fixated on solving it using the errorClass even for older versions of Spark.

jlowe replied:
Yeah the shim logic should almost always be "do what this version of Spark does for this" whether that's using an error class abstraction or throwing a raw AnalysisException with a string message.

razajafri replied:
OK. I think I got it right this time. This will be painful to review PTAL

jlowe commented on sql-plugin/src/main/spark311/scala/org/apache/spark/sql/rapids/shims/RapidsQueryErrorUtils.scala:
```diff
@@ -0,0 +1,139 @@
+trait RapidsQueryErrorUtils {
+
+  def createTableAsSelectWithNonEmptyDirectoryError(tablePath: String, conf: String): Throwable = {
+    new AnalysisException(s"CREATE-TABLE-AS-SELECT cannot create table with location to a " +
```
Many (all?) of these should be leveraging either QueryExecutionErrors or QueryCompilationErrors rather than throwing raw analysis exceptions on the Spark versions that support those classes, just like the Spark 4 version of this shim does.  Not sure it makes sense to isolate these as a trait from the other methods in RapidsErrorUtils.

razajafri replied:
I have moved the shim to Spark 320 which is  the version when the relevant `QueryCompilationErrors` were introduced

razajafri commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala:
```diff
@@ -121,7 +122,7 @@ case class GpuInsertIntoHadoopFsRelationCommand(
       (mode, pathExists) match {
         case (SaveMode.ErrorIfExists, true) =>
-          throw new AnalysisException(s"path $qualifiedOutputPath already exists.")
+          TrampolineUtil.throwAnalysisException(s"path $qualifiedOutputPath already exists.")
```
I saw that and decided not to use it because it's adding `Set mode as "overwrite" to overwrite the existing path.` which we didn't have in our original message.

jlowe replied:
But that extra message is applicable here, and we would want that.  The message was updated in Spark and we didn't update accordingly.

razajafri commented on sql-plugin/src/main/spark340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtilsFor340Plus.scala:
```diff
@@ -0,0 +1,95 @@
+/*
```
Renamed the old `RapidsErrorUtils`

razajafri commented:
I have addressed all your concerns I think PTAL. 
The only thing that you may have questions about is that I have introduced the RapidsAnalysisException which will be thrown even in older versions of Spark instead of Raw AnalysisException

razajafri commented:
build

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuInsertIntoHadoopFsRelationCommand.scala:
```diff
@@ -121,7 +121,7 @@ case class GpuInsertIntoHadoopFsRelationCommand(
       (mode, pathExists) match {
         case (SaveMode.ErrorIfExists, true) =>
-          throw new AnalysisException(s"path $qualifiedOutputPath already exists.")
+          throw RapidsErrorUtils.dataPathNotExistError(qualifiedOutputPath.toString)
```
This is the wrong error message to throw for this error.  The error message will state that it does _not_ exist, but instead it _does_exist, and that's the problem.  RapidsQueryErrorUtils.outputPathAlreadyExistsError would be more appropriate.  Note that the error being thrown here in Spark 4.0.0 is via QueryExecutionErrors.saveModeUnsupportedError.

razajafri replied:
Thanks. 

> Note that the error being thrown here in Spark 4.0.0 is via QueryExecutionErrors.saveModeUnsupportedError.

I am not sure what you mean here.

jlowe replied:
Just noting that there's a new interface in Spark 4.0.0 for throwing this type of error, see https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala#L139.  I think it's probably fine to use the outputPathAlreadyExistsError, but technically Spark 4.0.0 is using a different error utility method here.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuRandomExpressions.scala:
```diff
@@ -52,7 +52,7 @@ case class GpuRand(child: Expression) extends ShimUnaryExpression with GpuExpres
     case GpuLiteral(s, IntegerType) => s.asInstanceOf[Int]
     case GpuLiteral(s, LongType) => s.asInstanceOf[Long]
-    case _ => throw new AnalysisException(
+    case _ => TrampolineUtil.throwAnalysisException(
```
Why do we have a Trampoline method for throwing an AnalysisException _and_ a RapidsAnalysisException?  Seems like we should only have one of these, not both (probably just the latter).

mythrocks replied:
Looks like this was resolved by switching to RapidsAnalysisException.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala:
```diff
@@ -241,4 +241,11 @@ object TrampolineUtil {
 }
+
+/**
+ * This class is to only be used when there is no equivalent helper method in
+ * Spark to throw an AnalysisException.
```
```suggestion
 * This class is to only be used to throw errors specific to the
 * RAPIDS Accelerator or errors mirroring Spark where a raw
 * AnalysisException is thrown directly rather than via an error
 * utility class (this should be rare).
```

jlowe commented on sql-plugin/src/main/spark332db/scala/org/apache/spark/sql/rapids/GpuFileFormatWriter.scala:
```diff
@@ -61,7 +62,7 @@ object GpuFileFormatWriter extends Logging {
     schema.foreach { field =>
       if (!format.supportDataType(field.dataType)) {
-        throw new AnalysisException(
+        TrampolineUtil.throwAnalysisException(
```
Should use RapidsAnalysisException here.

jlowe commented on sql-plugin/src/main/spark400/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:
```diff
@@ -0,0 +1,31 @@
+  override def dynamicPartitionParentError: Throwable = {
+    /** This exception doesn't have a helper method so the errorClass has to be hardcoded */
+    throw new AnalysisException(
+      errorClass = "_LEGACY_ERROR_TEMP_3079",
```
I'm worried about this changing out from underneath us and then the error doesn't work at all.  The name itself implies this is temporary and will be changing, yet the compiler won't catch when that occurs.  It might be better to throw an error message directly here for now rather than rely on this tenuous connection to Spark's error message.

razajafri replied:
I have replaced this with the current way of throwing `ErrorMsg.PARTITION_DYN_STA_ORDER.getMsg`.

razajafri commented:
I have addressed your comments I think. 
* Renamed the method in `TrampolineUtil.throwAnalysisException()` to `TrampolineUtil.throwRapidsAnalysisException` 
* Throwing the ErrorMsg.PARTITION_DYN_STA_ORDER.getMsg in dynamicPartitionParentError instead of the errorClass
* Removed the `RapidsErrorUtilsFor340PlusShims`

razajafri commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/TrampolineUtil.scala:
```diff
@@ -157,8 +157,14 @@ object TrampolineUtil {
+   * AnalysisException is thrown directly rather than via an error
+   * utility class (this should be rare).
+   */
+  def throwRapidsAnalysisException(msg: String) = throw new RapidsAnalysisException(msg)
```
We don't need a `TrampolineUtil` method, this was a remnant of throwing the AnalysisException.

razajafri replied:
Let me get rid of that. Please let me know if there are any other concerns

razajafri commented:
build

mythrocks reviewed (approved):
I have verified that all of @jlowe's concerns were addressed here.

LGTM!
